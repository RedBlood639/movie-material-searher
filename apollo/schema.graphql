enum Media {
  movie
  tv
}

type UpcomingResults {
  id: Int!
  media_type: Media!
  poster_path: String
  release_date: String!
  title: String!
}

type Upcoming {
  page: Int!
  results: [UpcomingResults!]!
  total_pages: Int!
}

type NowPlayingResults {
  id: Int!
  media_type: Media!
  poster_path: String
  title: String!
  vote_average: Float!
}

type NowPlaying {
  page: Int!
  results: [NowPlayingResults!]!
  total_pages: Int!
}

type MoviesSearchResults {
  id: Int!
  media_type: Media!
  title: String!
}

type MoviesSearch {
  cursor: Int
  hasMore: Boolean
  page: Int!
  results: [MoviesSearchResults!]!
  total_pages: Int!
}

type SimilarResults {
  id: Int!
  media_type: Media!
  poster_path: String
  release_date: String!
  title: String!
}

type SimilarMovies {
  cursor: Int
  hasMore: Boolean
  results: [SimilarResults!]!
}

type MovieInfo {
  backdrop_path: String
  budget: String!
  id: Int!
  media_type: Media!
  overview: String
  poster_path: String
  release_date: String!
  revenue: String!
  similar: SimilarMovies!
  title: String!
  vote_average: Float!
}

type TvShowInfo {
  backdrop_path: String
  id: Int!
  media_type: Media!
  number_of_episodes: Int!
  number_of_seasons: Int!
  overview: String
  poster_path: String
  release_date: String!
  similar: SimilarMovies!
  title: String!
  vote_average: Float!
}

type Query {
  nowPlaying(page: Int = 1): NowPlaying!
  movieInfo(id: ID!, cursor: Int, pageSize: Int = 4): MovieInfo!
  moviesSearch(query: String!, cursor: Int, pageSize: Int = 8): MoviesSearch!
  tvShowInfo(id: ID!, cursor: Int, pageSize: Int = 4): TvShowInfo!
  upcoming(page: Int = 1): Upcoming!
}
