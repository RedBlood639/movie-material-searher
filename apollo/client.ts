import {
  ApolloClient,
  InMemoryCache,
  makeVar,
  NormalizedCacheObject,
} from '@apollo/client'
import merge from 'deepmerge'
import isEqual from 'lodash.isequal'
import { useMemo } from 'react'

export type Favorite = {
  id: number
  media_type: string
  poster_path?: string | null
  title: string
}

export const favoritesVar = makeVar<Favorite[]>([])
export const prefersDarkModeVar = makeVar(false)

let apolloClient: ApolloClient<NormalizedCacheObject>

function createApolloClient() {
  return new ApolloClient({
    cache: new InMemoryCache(),
    ssrMode: typeof window === 'undefined',
    uri: process.env.NEXT_PUBLIC_SERVER_URL,
  })
}

export function initializeApollo(initialState: {} | null = null) {
  const internalApolloClient = apolloClient ?? createApolloClient()

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = internalApolloClient.extract()

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = merge(initialState, existingCache, {
      // combine arrays using object equality (like in sets)
      arrayMerge: (destinationArray, sourceArray) => [
        ...sourceArray,
        ...destinationArray.filter(d => sourceArray.every(s => !isEqual(d, s))),
      ],
    })

    // Restore the cache with the merged data
    internalApolloClient.cache.restore(data)
  }
  // For SSG and SSR always create a new Apollo Client
  if (typeof window === 'undefined') return internalApolloClient
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = internalApolloClient

  return internalApolloClient
}

const APOLLO_STATE_PROP_NAME = '__APOLLO_STATE__'

export function addApolloState(
  client: ApolloClient<NormalizedCacheObject>,
  pageProps: any,
) {
  if (pageProps?.props) {
    // eslint-disable-next-line no-param-reassign
    pageProps.props[APOLLO_STATE_PROP_NAME] = client.cache.extract()
  }

  return pageProps
}

export function useApollo(pageProps: any) {
  const state = pageProps[APOLLO_STATE_PROP_NAME]
  const store = useMemo(() => initializeApollo(state), [state])
  return store
}
