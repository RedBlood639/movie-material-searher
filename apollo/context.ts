import MoviesAPI from './datasource'

export type Context = {
  dataSources: {
    moviesAPI: MoviesAPI
  }
}
