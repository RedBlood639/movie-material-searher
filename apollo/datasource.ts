/* eslint-disable class-methods-use-this */
import { RequestOptions, RESTDataSource } from 'apollo-datasource-rest'
import dayjs from 'dayjs'
import numeral from 'numeral'

import {
  Media,
  MovieInfo,
  MoviesSearch,
  NowPlaying,
  TvShowInfo,
  Upcoming,
} from './__generated__'
import {
  MockMovieInfoResponse,
  MockMoviesSearchResponse,
  MockNowPlayingResponse,
  MockTvShowInfoResponse,
  MockUpcomingResponse,
} from './__responses__/raw-responses'

export default class MoviesAPI extends RESTDataSource {
  constructor() {
    super()
    this.baseURL = 'https://api.themoviedb.org/3'
  }

  protected willSendRequest(request: RequestOptions) {
    request.params.set('api_key', process.env.MOVIE_API_KEY as string)
  }

  private attachPoster(path: string, size = 200) {
    return path ? `https://image.tmdb.org/t/p/w${size}${path}` : null
  }

  private moviesUpcomingReducer(movies: MockUpcomingResponse): Upcoming {
    return {
      page: movies.page,
      results: movies.results.map(movie => ({
        id: movie.id,
        media_type: Media.Movie,
        poster_path: this.attachPoster(movie.poster_path),
        release_date: dayjs(movie.release_date).format('DD.MM.YYYY'),
        title: movie.title,
      })),
      total_pages: movies.total_pages,
    }
  }

  private moviesNowPlayingReducer(movies: MockNowPlayingResponse): NowPlaying {
    return {
      page: movies.page,
      results: movies.results.map(movie => ({
        id: movie.id,
        media_type: Media.Movie,
        poster_path: this.attachPoster(movie.poster_path),
        title: movie.title,
        vote_average: movie.vote_average,
      })),
      total_pages: movies.total_pages,
    }
  }

  private moviesSearchReducer(movies: MockMoviesSearchResponse): MoviesSearch {
    return {
      page: movies.page,
      results: movies.results
        .filter(movie => movie.media_type !== 'person')
        .map(movie => ({
          id: movie.id,
          media_type: movie.media_type as Media,
          title: movie?.title || movie?.name || '',
        })),
      total_pages: movies.total_pages,
    }
  }

  private movieInfoReducer(movie: MockMovieInfoResponse): MovieInfo {
    return {
      backdrop_path: this.attachPoster(movie.backdrop_path, 500),
      budget: numeral(movie.budget).format('$0,00'),
      id: movie.id,
      media_type: Media.Movie,
      overview: movie.overview,
      poster_path: this.attachPoster(movie.poster_path),
      release_date: movie.release_date
        ? dayjs(movie.release_date).format('DD MMMM YYYY')
        : '',
      revenue: numeral(movie.revenue).format('$0,00'),
      similar: {
        results: movie.similar.results.map(similarMovie => ({
          id: similarMovie.id,
          media_type: Media.Movie,
          poster_path: this.attachPoster(similarMovie.poster_path),
          release_date: similarMovie.release_date
            ? dayjs(similarMovie.release_date).format('YYYY')
            : '',
          title: similarMovie.title,
        })),
      },
      title: movie.title,
      vote_average: movie.vote_average,
    }
  }

  private tvShowInfoReducer(tv: MockTvShowInfoResponse): TvShowInfo {
    return {
      backdrop_path: this.attachPoster(tv.backdrop_path, 500),
      id: tv.id,
      media_type: Media.Tv,
      number_of_episodes: tv.number_of_episodes,
      number_of_seasons: tv.number_of_seasons,
      overview: tv.overview,
      poster_path: this.attachPoster(tv.poster_path),
      release_date: tv.first_air_date
        ? dayjs(tv.first_air_date).format('DD MMMM YYYY')
        : '',
      similar: {
        results: tv.similar.results.map(similarTvShow => ({
          id: similarTvShow.id,
          media_type: Media.Tv,
          poster_path: this.attachPoster(similarTvShow.poster_path),
          release_date: similarTvShow.first_air_date
            ? dayjs(similarTvShow.first_air_date).format('YYYY')
            : '',
          title: similarTvShow.name,
        })),
      },
      title: tv.name,
      vote_average: tv.vote_average,
    }
  }

  async getUpcoming(page: number) {
    const data: MockUpcomingResponse = await this.get('/movie/upcoming', {
      page,
    })
    return this.moviesUpcomingReducer(data)
  }

  async getNowPlaying(page: number) {
    const data: MockNowPlayingResponse = await this.get('/movie/now_playing', {
      page,
    })
    return this.moviesNowPlayingReducer(data)
  }

  async getMoviesSearch(query: string, page: number) {
    const data: MockMoviesSearchResponse = await this.get('/search/multi', {
      page,
      query,
    })
    return this.moviesSearchReducer(data)
  }

  async getMovieInfo(id: string) {
    const data: MockMovieInfoResponse = await this.get(`/movie/${id}`, {
      append_to_response: 'similar',
    })
    return this.movieInfoReducer(data)
  }

  async getTvShowInfo(id: string) {
    const data: MockTvShowInfoResponse = await this.get(`/tv/${id}`, {
      append_to_response: 'similar',
    })
    return this.tvShowInfoReducer(data)
  }
}
