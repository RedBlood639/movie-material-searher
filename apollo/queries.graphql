query GetMovieInfo($id: ID!, $pageSize: Int) {
  movieInfo(id: $id, pageSize: $pageSize) {
    backdrop_path
    budget
    id
    media_type
    overview
    poster_path
    release_date
    revenue
    similar {
      results {
        id
        media_type
        poster_path
        release_date
        title
      }
    }
    title
    vote_average
  }
}

query GetTvShowInfo($id: ID!, $pageSize: Int) {
  tvShowInfo(id: $id, pageSize: $pageSize) {
    backdrop_path
    id
    media_type
    number_of_episodes
    number_of_seasons
    overview
    poster_path
    release_date
    similar {
      results {
        id
        media_type
        poster_path
        release_date
        title
      }
    }
    title
    vote_average
  }
}

query GetMovies {
  nowPlaying {
    page
    results {
      id
      media_type
      poster_path
      title
      vote_average
    }
    total_pages
  }
  upcoming {
    page
    results {
      id
      media_type
      poster_path
      release_date
      title
    }
    total_pages
  }
}

query GetUpcoming($page: Int) {
  upcoming(page: $page) {
    page
    results {
      id
      media_type
      poster_path
      release_date
      title
    }
    total_pages
  }
}

query GetNowPlaying($page: Int) {
  nowPlaying(page: $page) {
    page
    results {
      media_type
      id
      poster_path
      title
      vote_average
    }
    total_pages
  }
}

query GetMoviesSearch($query: String!, $pageSize: Int) {
  moviesSearch(query: $query, pageSize: $pageSize) {
    results {
      id
      media_type
      title
    }
  }
}
