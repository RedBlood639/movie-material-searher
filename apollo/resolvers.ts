import { MoviesSearch, Resolvers, SimilarMovies } from './__generated__'

function paginateResults<T extends { results: { [key: string]: any } }>(
  data: T,
  size: number,
  cursor?: number | null,
) {
  let results: []

  if (!cursor) {
    results = data.results.slice(0, size)
  } else {
    results = data.results.slice(0, cursor + size)
  }

  return {
    ...data,
    cursor: results.length,
    hasMore: results.length !== data.results.length,
    results,
  }
}

const resolvers: Resolvers = {
  Query: {
    async movieInfo(_root, { cursor, id, pageSize }, { dataSources }) {
      try {
        const data = await dataSources.moviesAPI.getMovieInfo(id)
        return {
          ...data,
          similar: paginateResults<SimilarMovies>(
            data.similar,
            pageSize,
            cursor,
          ),
        }
      } catch (error) {
        throw new Error(error)
      }
    },
    async moviesSearch(_root, { cursor, pageSize, query }, { dataSources }) {
      const fetchMovies = async (page = 1): Promise<MoviesSearch> => {
        const data = await dataSources.moviesAPI.getMoviesSearch(query, page)

        if (!data.results.length && data.total_pages > 1) {
          return fetchMovies(data.page + 1)
        }

        return data
      }

      try {
        const data = await fetchMovies()
        return paginateResults<MoviesSearch>(data, pageSize, cursor)
      } catch (error) {
        throw new Error(error)
      }
    },
    async nowPlaying(_root, { page }, { dataSources }) {
      try {
        const data = await dataSources.moviesAPI.getNowPlaying(page)

        if (!data.results.length || data.page !== +page) {
          throw new Error('404 Not found')
        }

        return data
      } catch (error) {
        throw new Error(error)
      }
    },
    async tvShowInfo(_root, { cursor, id, pageSize }, { dataSources }) {
      try {
        const data = await dataSources.moviesAPI.getTvShowInfo(id)
        return {
          ...data,
          similar: paginateResults<SimilarMovies>(
            data.similar,
            pageSize,
            cursor,
          ),
        }
      } catch (error) {
        throw new Error(error)
      }
    },
    async upcoming(_root, { page }, { dataSources }) {
      try {
        const data = await dataSources.moviesAPI.getUpcoming(page)

        if (!data.results.length || data.page !== +page) {
          throw new Error('404 Not found')
        }

        return data
      } catch (error) {
        throw new Error(error)
      }
    },
  },
}

export default resolvers
