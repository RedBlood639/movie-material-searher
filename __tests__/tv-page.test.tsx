import { GetTvShowInfoDocument } from '@/apollo/__generated__'
import { renderApollo, screen } from '@/lib/setupTests'
import TvPage from '@/pages/tv/[id]'

const mocks = [
  {
    request: {
      query: GetTvShowInfoDocument,
      variables: {
        id: '1',
      },
    },
    result: {
      data: {
        tvShowInfo: {
          __typename: 'TvShowInfo',
          backdrop_path: null,
          id: 1,
          media_type: 'tv',
          number_of_episodes: 30,
          number_of_seasons: 5,
          overview: 'overview',
          poster_path: null,
          release_date: '2002',
          similar: {
            results: [
              {
                __typename: 'SimilarResults',
                id: 2,
                media_type: 'tv',
                poster_path: null,
                release_date: '2002',
                title: 'rendered similar',
              },
            ],
          },
          title: 'rendered tv',
          vote_average: 10,
        },
      },
    },
  },
]

describe('tv page', () => {
  it('should fetch tv by id', async () => {
    const { baseElement } = renderApollo(<TvPage id="1" />, {
      addTypeName: true,
      mocks,
    })

    expect(await screen.findByText('rendered tv')).toBeInTheDocument()
    expect(await screen.findByText('rendered similar')).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('should render error state', async () => {
    const mock = [
      {
        error: new Error('an error has occurred'),
        request: {
          query: GetTvShowInfoDocument,
          variables: {
            id: '1',
          },
        },
      },
    ]

    renderApollo(<TvPage id="1" />, { mocks: mock })

    expect(
      await screen.findByText(/an error has occurred/i),
    ).toBeInTheDocument()
  })
})
