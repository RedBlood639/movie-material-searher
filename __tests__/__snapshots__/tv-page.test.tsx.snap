// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tv page should fetch tv by id 1`] = `
<body>
  <div>
    <div
      style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation10 MuiCard-root eeexpct5 css-zdrws2-MuiPaper-root-MuiCard-root-StyledCard"
      >
        <div
          class="css-kqif1p-ImageWrapper eeexpct4"
        >
          <div
            style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
          >
            <img
              alt="rendered tv"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
            />
          </div>
        </div>
        <div
          class="MuiCardContent-root eeexpct3 css-u5tic9-MuiCardContent-root-StyledCardContent"
        >
          <div
            class="css-bgyst3-TitleWrapper eeexpct0"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
            >
              rendered tv
            </h5>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root eeexpct1 css-1ob3qkx-MuiButtonBase-root-MuiButton-root-StyledButton"
              tabindex="0"
            >
              <span
                class="MuiButton-label css-8xplcm-MuiButton-label"
              >
                Add to favorites
              </span>
            </button>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph css-z2eky3-MuiTypography-root"
          >
            overview
          </p>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <p
            class="MuiTypography-root MuiTypography-body1 eeexpct2 css-lj0jjw-MuiTypography-root-StyledTypography"
          >
            <b>
              Seasons:
            </b>
             
            5
          </p>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <p
            class="MuiTypography-root MuiTypography-body1 eeexpct2 css-lj0jjw-MuiTypography-root-StyledTypography"
          >
            <b>
              Episodes:
            </b>
             
            30
          </p>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <p
            class="MuiTypography-root MuiTypography-body1 eeexpct2 css-lj0jjw-MuiTypography-root-StyledTypography"
          >
            <b>
              Rating:
            </b>
             
            10
          </p>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <p
            class="MuiTypography-root MuiTypography-body1 eeexpct2 css-lj0jjw-MuiTypography-root-StyledTypography"
          >
            <b>
              Release Date:
            </b>
             
            2002
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation10 efhkdvm1 css-103l67h-MuiPaper-root-Container"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter MuiTypography-gutterBottom css-1trk376-MuiTypography-root"
        >
          Similar Movies
        </h4>
        <div
          class="MuiGrid-root MuiGrid-container css-15fev7r-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-3 efhkdvm0 css-11ine10-MuiGrid-root-Wrapper"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ejzktm53 css-18ogkrw-MuiPaper-root-MuiCard-root-CardContainer"
            >
              <button
                class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
                tabindex="0"
                type="button"
              >
                <div
                  style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                >
                  <div
                    style="box-sizing: border-box; display: block; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      role="presentation"
                      src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjAwIiBoZWlnaHQ9IjMwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                      style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                    />
                  </div>
                  <img
                    alt="rendered similar"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
                  />
                </div>
                <div
                  class="MuiCardContent-root ejzktm52 css-dzh1mx-MuiCardContent-root-StyledCardContent"
                >
                  <p
                    class="MuiTypography-root MuiTypography-inherit ejzktm50 css-eh8d4t-MuiTypography-root-StyledTypography"
                  >
                    rendered similar
                  </p>
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                  >
                    2002
                  </p>
                </div>
                <span
                  class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
